### PR TITLE
Port SitemapExtractor from `CC-MRJob` to `CC-PySpark`

### PR DESCRIPTION
**Pull request by damian0815**
Port [cc-mrjob/sitemaps_from_robotstxt.py](https://github.com/commoncrawl/cc-mrjob/blob/master/sitemaps_from_robotstxt.py).

* [x] Basic functionality unit tests
* [x] warcio implementation
  * [x] Validate output is identical to MRJob output with "test" robotstxt in MRJob repo
  * [x] Validate on recent full-scale crawl output
* [x] fastwarc implementation
* [x] unit test to validate text encoding edge cases and validity (currently all test cases are completely valid utf8)
* [x] check output works with `crawl-tools/server/seed/sitemaps/sitemaps_robotstxt.py`

### CODE REVIEW

File: test/test_sitemaps_from_robotstxt.py [hunk 1]
```diff
@@ -0,0 +1,576 @@
...
+
+Sitemap: http://example.com/sitemap.xml
+Sitemap: http://example.com/sitemap2.xml
+""".encode('utf-8')
+    )
+
+    job = SitemapExtractorJob()
+    job.init_accumulators(session=spark)
+    results = list(job.process_record(record))
+    assert len(results) == 0
```
Comment (sebastian-nagel) [inline test/test_sitemaps_from_robotstxt.py]: Shouldn't it be 2 because both sitemap URLs are valid? If the WARC-Target-URI does not validate, it only effects the cross-submit list. But this will hardly happen. In all crawls since Aug 2018 the WARC-Target-URI is validated as URI.

File: sitemaps_from_robotstxt.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+                self.sitemap_urls_found.add(1)
+                n_sitemaps += 1
+                try:
+                    sitemap_url = sitemap_url.decode("utf-8", "strict")
+                except UnicodeDecodeError as e:
+                    self.get_logger().warn(f'Invalid encoding of sitemap URL {sitemap_url}: {repr(e)}')
+                    self.sitemap_url_invalid_encoding.add(1)
+                    continue
+
+                if not self._is_valid_url(sitemap_url, label_for_log='sitemap'):
```
Comment (sebastian-nagel) [inline sitemaps_from_robotstxt.py]: 1. Does validation make sense before resolving the URL given the base URL (WARC-Target-URI)?
2. URL validation should be optional because
   - It's extra work (which is done anyway if you fetch the sitemaps)
   - We need to ensure that the [URL validator](https://github.com/python-validators/validators/blob/master/src/validators/url.py) does the same as the URL normalization and validation implemented in the crawler. This is a huge task, out-of-scope.

**Review by sebastian-nagel — CHANGES_REQUESTED** [review]
First pass. I'll continue with testing. But several points need a discussion.

File: sparkcc.py [hunk 3]
```diff
@@ -34,6 +35,8 @@ class CCSparkJob(object):
         StructField("val", LongType(), True)
     ])
 
+    merge_method: Literal['reduce_by_key', 'reduce_group_by_key'] = 'reduce_by_key'
```
[the remark this answers is not in the document]
Comment (damian0815) [reply sparkcc.py]: Good question! I initially went with what you were suggesting but then deferred to preserving the `mrjob` structure which does it like this (first pass produces sitemap/host pairs, second pass groups) - assuming it was split like this for performance reasons. If this isn't a concern then I agree it would be better to do it all in `run_job`.

File: test/test_sitemaps_from_robotstxt.py [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+def test_sitemap_url_invalid_encoding_mixed_bytes(spark):
+    """ Test incorrectly encoded sitemap URL - mixed UTF-8 invalid bytes """
+    # The byte sequence \xff\xfe is not valid UTF-8
+    record = make_robots_txt_record(
+        warc_target_uri='http://example.com/robots.txt',
+        response_bytes=b"""User-agent: *
+Disallow: /search
+
+Sitemap: http://example.com/good_sitemap.xml
+Sitemap: http://example.com/bad\xff\xfe_sitemap.xml
```
[the remark this answers is not in the document]
Comment (damian0815) [reply test/test_sitemaps_from_robotstxt.py]: Right - the point is for it to be improperly encoded so that the error reporting logic can get tested. My intention with these tests was to write test code that triggered all the different error reporting mechanisms on the old mrjob (see `init_accumulators`. If it's unnecessary then of course we can remove them, but that would leave us with error-handling logic which doesn't have test coverage.

File: sitemaps_from_robotstxt.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
+import json
+import re
+from typing import Optional
+from urllib.parse import urlparse, urljoin
+
+import validators
```
Comment (sebastian-nagel) [inline sitemaps_from_robotstxt.py]: Not required anymore.

File: test/test_sitemaps_from_robotstxt.py [hunk 6]
```diff
@@ -0,0 +1,553 @@
...
+    assert results[0][0] == 'http://3535.ru/sitemap_000.xml'
+    assert results[0][1] == ['177.52.3535.ru']
+
+
+
+
+def test_host_accumulation_same_host(spark):
+    """
+    Test accumulation of hosts when sitemap url host and robots.txt url host match
+    Requires test/ on PYTHONPATH so utils._process_jobs can be imported
```
Comment (sebastian-nagel) [inline test/test_sitemaps_from_robotstxt.py]: A 4-line function could be inlined to avoid the import error.

However, `test/` is required on the path also to load `test_sitemaps_from_robotstxt` which is required for Spark serialization. So, you need to run it like
```
PYTHONPATH=$PYTHONPATH:./test python -m pytest test -v
```

File: sitemaps_from_robotstxt.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+                try:
+                    sitemap_url = sitemap_url.decode("utf-8", "strict")
+                except UnicodeDecodeError as e:
+                    self.get_logger().warn(f'Invalid encoding of sitemap URL {sitemap_url}: {repr(e)}')
+                    self.sitemap_url_invalid_encoding.add(1)
+                    continue
+
+                if robots_txt_url is None:
+                    # first sitemap found: set base URL and get host from URL
+                    robots_txt_url = record.rec_headers['WARC-Target-URI']
```
Comment (sebastian-nagel) [inline sitemaps_from_robotstxt.py]: This is not compatible with FastWARC, should be `self.get_warc_header(record, 'WARC-Target-URI')`

File: sitemaps_from_robotstxt_fastwarc.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+class SitemapExtractorFastWarcJob(SitemapExtractorJob, CCFastWarcSparkJob):
+    """Extract sitemap URLs (http://www.sitemaps.org/) from robots.txt WARC files
+       using FastWARC to parse WARC files."""
+
+    name = "SitemapExtractorFastWarc"
+
+    # process only WARC response and metadata (including WAT) records
+    fastwarc_record_filter = WarcRecordType.response
+
+    # process_record is implemented by SitemapExtractorJob
```
[the remark this answers is not in the document]
Comment (sebastian-nagel) [reply sitemaps_from_robotstxt_fastwarc.py]: Please, also run the job to verify that there are no errors.

**Review by sebastian-nagel — CHANGES_REQUESTED** [review]
Great! Some minor things remain to do.

Tested on my local machine:
- successfully run the unit tests using both the pyspark module or the installed Spark. For the latter, it's required to set `PYTHONPATH=$PWD/test:$(ls $SPARK_HOME/python/lib/py4j-*-src.zip):$SPARK_HOME/python:$PYTHONPATH`
- successfully tested `sitemaps_from_robotstxt.py`. Output looks good. On a small test sample, there are no differences in the number of extracted sitemap URLs with the cc-mrjob implementation.
- failed to run `sitemaps_from_robotstxt_fastwarc.py`

File: sitemaps_from_robotstxt.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+                    try:
+                        robots_txt_host = urlparse(robots_txt_url).netloc.lower().lstrip('.')
+                    except Exception as e1:
+                        try:
```
[the remark this answers is not in the document]
Comment (damian0815) [reply sitemaps_from_robotstxt.py]: I inherited this from the cc-mrjob code:
https://github.com/commoncrawl/cc-mrjob/blob/1d1cf0ed6c7056f209b5ab97795d8afbe724eb30/sitemaps_from_robotstxt.py#L58

File: sitemaps_from_robotstxt.py (hunk 5 above)
Comment (damian0815) [reply sitemaps_from_robotstxt.py]: Fixed

File: sitemaps_from_robotstxt_fastwarc.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (damian0815) [reply sitemaps_from_robotstxt_fastwarc.py]: fixed

**Review by sebastian-nagel — APPROVED** [review]
Excellent, @damian0815!

Unit tests pass, successfully run both versions (warcio and fastwarc) locally.

I'll also run the job on a real cluster later today and merge the PR if this test passes as well.

Would you mind to squash the commits to a small and meaningful number? But I can do it when merging as well. Thanks!

**Comment by damian0815** [comment]
I think it's easier/cleaner if you squash when merging?

**Review by sebastian-nagel — APPROVED** [review]
I've successfully run `sitemaps_from_robotstxt_fastwarc.py` on 5% of the robots.txt captures of CC-MAIN-2025-43 using a single-node Hadoop cluster (Spark on YARN):
```
13:57:14.391 [Thread-5] INFO  SitemapExtractorFastWarc - WARC/WAT/WET input files processed = 5000
13:57:14.394 [Thread-5] INFO  SitemapExtractorFastWarc - WARC/WAT/WET input files failed = 0
13:57:14.396 [Thread-5] INFO  SitemapExtractorFastWarc - WARC/WAT/WET records processed = 4444146
13:57:14.398 [Thread-5] INFO  SitemapExtractorFastWarc - robots.txt successfully parsed = 4444146
13:57:14.401 [Thread-5] INFO  SitemapExtractorFastWarc - sitemap urls found = 3949561
13:57:14.403 [Thread-5] INFO  SitemapExtractorFastWarc - sitemap urls with invalid utf-8 encoding = 76
13:57:14.405 [Thread-5] INFO  SitemapExtractorFastWarc - robots.txt announcing at least 1 sitemap = 1885795
13:57:14.408 [Thread-5] INFO  SitemapExtractorFastWarc - robots.txt with more than 50 sitemaps = 3547
```

I'll merge the code. Thanks, @damian0815!
